pkg/lib/cockpit-components-file-autocomplete.jsx

```diff
@@ -0,0 +1,246 @@
+            if (ap > bp)
+                return 1;
+            else if (ap < bp)
+                return -1;
+            else
+                return 0;
+        });
+
+        this.setState({
+            displayFiles: result,
+            directoryFiles: result,
+            error: error,
+        });
+    },
+
+    filterFiles: function(value) {
+        var inputValue = value.trim().toLowerCase();
+        var dirLength = this.state.directory.length;
+        var matches = [];
+        var inputLength;
+
+        inputValue = inputValue.slice(dirLength);
+        inputLength = inputValue.length;
+
+        if (this.state.directoryFiles !== null) {
+            matches = this.state.directoryFiles.filter(function (v) {
+                return v.path.toLowerCase().slice(0, inputLength) === inputValue;
+            });
+        }
+
+        this.setState({
+            displayFiles: matches,
+            open: true,
+        });
+    },
+
+    showAllOptions: function (ev) {
+        // only consider clicks with the primary button
+        if (ev && ev.button !== 0)
+            return;
+
+        this.setState({
+            open: !this.state.open,
+            displayFiles: this.state.directoryFiles || [],
+        });
+    },
+
+    selectItem: function (ev) {
+        // only consider clicks with the primary button
+        if (ev && ev.button !== 0)
+            return;
+
+        if (ev.target.tagName == 'A') {
+            var value = ev.target.innerText;
+            var directory = this.state.directory || "/";
+
+            if (directory.charAt(directory.length - 1) !== '/')
+                directory = directory + "/";
+
+            value = directory + value;
+            this.setState({
+                open: false,
+                value: value
+            });
+            this.updateIfDirectoryChanged(value);
+        }
+    },
+
+    renderError: function(error) {
+        return (
+            <li className="alert alert-warning">
+                {error}
+            </li>
+        );
+    },
+
+    render: function() {
+        var placeholder = this.props.placeholder || _("Path to file");
+        var controlClasses = "form-control-feedback ";
+        var classes = "input-group";
+        if (this.state.open)
+            classes += " open";
+
+        if (this.state.directoryFiles === null)
+            controlClasses += "spinner spinner-xs spinner-inline";
+        else
+            controlClasses += "caret";
+
+
+        var listItems, error;
+        if (this.state.error)
+            error = cockpit.format(cockpit.message(this.state.error));
+        else if (this.state.directoryFiles && this.state.displayFiles.length < 1)
+            error = _("No matching files found");
+
+        if (error) {
+            listItems = [this.renderError(error)];
+            classes += " error"
+        } else {
+            listItems = React.Children.map(this.state.displayFiles, function(file) {
```

Comment:
We could have shown those 5000 and elided the rest, but that could also be confusing... this is good.